WIP | Migrate to ROS 2

ruffsl opened the pull request:
* Update apparmor library for ROS2
* Update cpp and python examples
* Re-audit permission footprint for ROS programs

ruffsl committed:
Set ~/.ros path via ROS_HOME
instead of ROS_LOCAL

ruffsl committed:
Glob ROS_INSTALL over any ros distro

ruffsl committed:
Include ROS_INSTALL and ROS_HOME

ruffsl committed:
Default for python 3

ruffsl committed:
Prune unused node permissions

ruffsl committed:
Update Python node permissions

ruffsl committed:
Add cpp and python node examples

ruffsl committed:
Remove old ros1 demo example

ruffsl committed:
Tighten read permissions in example

ruffsl committed:
Update date

ruffsl committed:
Glob all .so like files

ruffsl committed:
Fix python example

ruffsl committed:
Use nameservice abstractions

ruffsl committed:
Remove trailing slash

ruffsl committed:
Update python expression

ruffsl committed:
Tidy up paths

ruffsl committed:
Allow listing the directory

ruffsl committed:
Expand permissions for ROS_HOME
such as locking and linking

mikaelarguedas commented on profiles/ros/python:
```diff
@@ -11,4 +11,28 @@
+@{PROC}/@{pid}/environ r,
+@{PROC}/@{pid}/fd/ r,
+
+# CHECKME: is `python3.[0-6]` too relaxed?
```
Python 3.5 (xenial uses 3.5.1) and higher should be enough

ruffsl replied:
Ok, I'll ratchet that up to  `python3.[5-6]` then.

mikaelarguedas commented on profiles/ros/base:
```diff
@@ -11,13 +11,15 @@
 
 # Allow us to receive signals from ros tools
-signal receive peer=ros/roslaunch,
+# signal receive peer=ros/roslaunch,
```
We should be able to remove the roslaunch example https://github.com/ros-infrastructure/apparmor_profiles/blob/4418fada82e47fca862211ebf0a4065b93de0dcc/examples/ros/roslaunch_talker_listener

Or to update it with a profile for `ros2 launch demo_nodes_cpp talker_listener.launch.py`

ruffsl replied:
Yea, I haven't touched the launch example files as I've yet to profile the permission requirements for ROS2's launch process. I'm also not sure of the AppArmor profile language as changed at all here, perhaps there alternate methods of scoping signal prevleage by now.

ruffsl commented on examples/ros/opt.ros.distro.lib.demo_nodes_cpp:
```diff
@@ -0,0 +1,12 @@
+#include <tunables/global>
+#include <tunables/ros>
+
+profile demo_nodes_cpp/talker @{ROS_INSTALL_LIB}/demo_nodes_cpp/talker {
```
Looking at this now, I wonder what has changed wrt. AppArmor as for the executable permission for executable path to no longer be necessary, as compared to previous profiles I've tested. Is this due to an upstream change in an abstraction I've imported?

ruffsl commented on profiles/tunables/ros.d/install:
```diff
@@ -1,17 +1,13 @@
 # directories are stored. Directories added here are appended to @{ROS_INSTALL}.
-# See ros/home for details.
-@{ROS_INSTALL}=/opt/ros/kinetic
+@{ROS_INSTALL}=/opt/ros/*
```
I've updated this profile library wholesale for ROS2, though I'm not sure there is a better why to separate privileges to ros2 versions without hardcoding the distro (which I think is rather fragel and tedious). We could template some things for auto generation during install time, but I'm not sure we need to overcomplicate things. Like what if multiple distros are installed simultaneously, should the debian packages just state they conflict (I'd guess so)? 

However, do folks think we should add the version to the paths and imports, e.g.

``` diff
#include <tunables/global>
- #include <tunables/ros>
+ #include <tunables/ros2>

- profile demo_nodes_cpp/talker @{ROS_INSTALL_LIB}/demo_nodes_cpp/talker {
-  #include <ros/base>
-  #include <ros/node>
+ profile demo_nodes_cpp/talker @{ROS2_INSTALL_LIB}/demo_nodes_cpp/talker {
+  #include <ros2/base>
+  #include <ros2/node>
}
```

ruffsl committed:
Spelling

ruffsl committed:
Add note about why nodes are trying to read OpenSSL Configuration File

```
Jun 12 11:46:55 dox kernel: [66614.402478] audit: type=1400 audit(1560365215.296:184): apparmor="DENIED" operation="open" profile="demo_nodes_cpp/talker" name="/etc/ssl/openssl.cnf" pid=19859 comm="talker" requested_mask="r" denied_mask="r" fsuid=1000 ouid=0
```

ruffsl committed:
Glob over python releases

ruffsl committed:
Restrict ROS_HOME access to logging

ruffsl committed:
Update example profile names
and add read and map permissions for profile transisioning

ruffsl committed:
Glob over python versions

ruffsl committed:
Clean up abstraction

ruffsl committed:
remove old ros launch profile

ruffsl committed:
Add development profile for ros2 cli

ruffsl committed:
Breakup node abstractions

ruffsl committed:
Update the python example

ruffsl committed:
Rename imports

ruffsl committed:
Fix python path

ruffsl committed:
Remove duplicate openssl rule

ruffsl committed:
Reorganize folder structure

ruffsl committed:
Fix python deny events

ruffsl committed:
Modularize python import

ruffsl committed:
Update readme with new examples

ruffsl committed:
Extend python deny rules for __pycache__

ruffsl committed:
Prune duplicate error message for readability

ruffsl committed:
Remove old profile
